### PR TITLE
Stop execution of chained thens after error

### PR DESCRIPTION
**Pull request by matthewrfindley**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ function buildThenable() {
   return {
     then: function(onFulfill, onReject) {
       try {
-        if (this.resolved) {
+        if (this.resolved && !this.rejected) {
```
Comment (astephenb) [inline index.js]: I'm curious how `resolve` and `rejected` end up being true at the same time?

**Comment by astephenb** [comment]
My question is not a blocker, just curiosity. This looks good! :+1:

**Comment by matthewrfindley** [comment]
I agree, I am curious as well, but only wanted to find a quick solution to the problem.

File: index.js (hunk 1 above)
Comment (amay) [inline index.js]: My guess would be that we call `resolves` in one `beforeEach` and then `rejects` in another. We really need to make our resolves/rejects able to be dependent on the arguments to the stub (ie. make `withArgs` work for stubbed promises). That would let us do conditional behavior based on args, which would move us to more of a stubbing pattern vs. mocking, and would probably eliminate scenarios where we do both of these.

**Comment by amay** [comment]
FYI: there is a strange scenario with then/catch chaining and promises that this made me think of.

Try this:
```javascript
Promise.resolve('foo').then(() => {throw new Error('error')}).catch((e) => console.log(e)).then(() => console.log('in second then'))
```
You will see that the catch logs the error, as expected, but then the "in second then" log is also show. The behavior is that unless the catch throws the error again, subsequent items chained on will start executing again normally.

This change would break that behavior, but I don't believe we are actually using that behavior anywhere (yet). And we are definitely relying on the behavior of a thrown error skipping execution of subsequent then blocks, so I'm totally on board with this change. I just wanted to point this out for future reference.

**Pull request merged**